### PR TITLE
Formalize Testing Rules

### PR DESCRIPTION
**Pull request by silv-io**
## Motivation

We recurrently experience stability issues in our CI pipeline. Many of the issues we encounter are avoidable and similar to each other. However, for those issues to be avoided we need common rules to adhere to.

## Changes

This PR slightly restructures the testing part of the contribution documentation and adds on a ruleset for stable testing pipelines. The ruleset is inspired by recent issues with pipeline stability and how those issues were resolved.

Further input is greatly appreciated! :)

### CODE REVIEW

File: docs/testing/README.md
Comment (alexrashed) [inline docs/testing/README.md:17]: R02 are common issues with our test pipeline, but we don't really have a clear solution yet.
And I am not really sure what R02a means exactly to be honest...

File: docs/testing/README.md
Comment (thrau) [inline docs/testing/README.md:16]: this one is not clear to me. could you elaborate/rephrase? why would i assume indefinitely available resources in a test? is there an example?

File: docs/testing/README.md
Comment (alexrashed) [inline docs/testing/README.md:15]: Maybe we should even directly file a PR marking it as skip? This would automatically inform code owners that a test is flaky 😛

File: docs/testing/README.md
Comment (thrau) [inline docs/testing/README.md:18]: does this just refer to error reporting?

**Review by alexrashed — COMMENTED** [review]
This is a great spike for clearer and cleaner testing guidelines and rules for both authors and reviewers! 🚀 
I added a few comments, but I think this should really be a PR where we have an initial spike used to kickstart a discussion, which then converges to the best possible outcome.
Which is why I think this PR should also see some attention from others (especially @dominikschubert, @HarshCasper, @thrau) in order to converge to a state where we have a clear set of guidelines which _help_ both authors of tests and reviewers of PRs containing these tests. Any input is highly appreciated!

File: docs/testing/README.md
Comment (thrau) [inline docs/testing/README.md:20]: perhaps we should rephrase this to not wait a _set_ amount of time in general, but rather use event-based notification mechanisms (like using SQS queues to wait for lambda results) where possible, and otherwise use polling with reasonable timeouts?
perhaps also adding that this is specifically for asynchronous operations?

**Review by thrau — COMMENTED** [review]
I think this is a very nice idea, and a great first iteration! I like the use of examples, and most of the rules are clear. I added comments where I think clarification is needed.

:100:

File: docs/testing/integration-tests/README.md
Comment (silv-io) [reply docs/testing/integration-tests/README.md:7]: I'm not sure I understand. Do you mean the "general rules" contain things that are more akin to "guiding principles" or the other way around, or both ways?

File: docs/testing/README.md
Comment (silv-io) [reply docs/testing/README.md:13]: I'm not sure every rule would fit into this scheme, but for those which do we can add this for sure. Makes them easier to be memorized as well

File: docs/testing/README.md
Comment (silv-io) [reply docs/testing/README.md:17]: In a failing test involving pods, it failed because some S3 bucket that got accessed was deleted. I'm not sure if it was a write or a read access.
But generally, if there is some test where something should be written to a bucket `x`, you can add a check beforehand if `x` exists, and if it doesn't just create it before using it.

Comment (silv-io) [reply docs/testing/README.md:17]: Also, I think manifesting mindfulness of these issues in the rules will lead to development of best practices for these problems in the future

File: docs/testing/README.md
Comment (silv-io) [reply docs/testing/README.md:15]: That is what I was thinking about concerning the "Inform" part. I will make this more explicit!

File: docs/testing/README.md
Comment (silv-io) [reply docs/testing/README.md:16]: We have tests which access remote resources. Some access our own resources, like tests surrounding pods, and some access some files (like libraries, container images) stored on servers which are not under our control.

File: docs/testing/README.md
Comment (silv-io) [reply docs/testing/README.md:18]: Yes, access errors should be clearly marked as such, and as Alex mentioned should lead to fast failures.
This helps avoiding needless debugging of our own code when the issue lies outside of it.

In the cases where this was relevant in the past, we actually did well here. But making it a general rule will help us keep it like this in the future. If this is too trivial, we can also remove it again, though :)

File: docs/testing/README.md
Comment (silv-io) [reply docs/testing/README.md:20]: This is what I was trying to say with that but couldn't properly express it concisely enough

File: docs/testing/README.md
Comment (dominikschubert) [reply docs/testing/README.md:18]: Still a bit confused about the wording around `access` here. Since we're for the most part in the context of AWS this feels like you're talking about IAM permissions. Maybe rewording this might clear things up? E.g. retrieving/using vs. accessing?

File: docs/testing/README.md
Comment (dominikschubert) [reply docs/testing/README.md:19]: So I assume this are things like: authenticated API access, retry logic, ... ? Anything else that would fit under here?

File: docs/testing/README.md
Comment (dominikschubert) [inline docs/testing/README.md:25]: Maybe generally adding the list of test markers + explanations here in the README might be quite nice as well!

**Review by alexrashed — COMMENTED** [review]
This is already looking really good! 🚀 
I think we can move forward after this round of reviews is processed and merge the PR 🥳

File: docs/testing/README.md
Comment (alexrashed) [inline docs/testing/README.md:292]: This seems to be a leftover.

**Review by alexrashed — APPROVED** [review]
Awesome! The cross-linking and the additional description is super helpful! Thanks for distilling these rules / guidelines!
As soon as the leftover (see comment) is removed, I think this should be merged! 🥳 
Let's make sure that we come back to these rules, explicitly reference them in reviews, and most importantly evolve them over time! 🚀

**Pull request merged**
